File: src/PYTHON/Termostato.c

// Set simulation parameters


struct _andersen_termostat {
	double Temp;
	double nu;
	double velSTD;
};
typedef struct _andersen_termostat a_termostat;

void Andersen_termostat(mdsys_t *sys, a_termostat *a_term)
{

	int i;

	double dt = sys->dt;        // Integration time
	double dt2 = dt*dt;        // Integration time, squared

	double Temp = a_term->Temp;         // Simulation temperature
	double nu = a_term->nu;           // Thermostat parameter - frequency of collisions with the heat bath

	a_term->velSTD = sqrt(Temp/sys->mass); // Thermostat parameter - standard deviation of the velocity
 
        for (i = 0; i < 3*sys->natoms; ++i) {
        sys->ekin += sys->vel[i]*sys->vel[i];
    }
	// Implement the Andersen thermostat
	for(i = 0; i < self->natoms, i++)
	{
	//	Test for collisions with the Andersen heat bath
        
		if (rand() < nu*dt)
		{
	              // If the particle collided with the bath, draw a new velocity
        	      // out of a normal distribution
        	      sys->vel[i] = randGauss(0,velSTD,nDim);
        	      sys->vel[i + sys->natoms] = randGauss(0,velSTD,nDim);
        	      sys->vel[i + (sys->natoms)*2] = randGauss(0,velSTD,nDim);
		}
	}
}	

File: src/PYTHON/termostato.py
import random as rand
import numpy as np

class Andersen_termostat(object):

    def __init__(self, temp, nu):
        self.temp = temp
	self.nu = nu
	

    def set_temp(self, mdsys_t):

        natoms = mdsys_t.natoms

        kboltz = 0.0019872067

       # velstd = (mdsys_t.mass / (kboltz * self.temp))**0.5

        velstd = 20

        for i in range(0, natoms):

            if rand.random() < (self.nu * mdsys_t.dt):

                mdsys_t.vel[i] = np.random.normal(0, velstd)
                mdsys_t.vel[i + natoms] = np.random.normal(0, velstd)
                mdsys_t.vel[i + 2*natoms] = np.random.normal(0, velstd)
         

